### DISCUSSION TITLE
self.parameters() access in pytorch lightning

### DISCUSSION THREAD

**Discussion by arjunsinghrathore** [Lightning-AI/pytorch-lightning#13394] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule]
I have noticed that if i call self.parameters() like this in the init of the pl.LightningModule class, then I'm able to access it :

`for param in self.parameters():
        print('param : ',param.weight)`

but if I call it similarly either in the training step or the optimizer step then self.parameters() does not return any param. Is there something special I need to do to rectify this?

**Comment by rohitgr7** [comment]
I don't think `param`has an attribute `weight`. Can you share a small reproducible example?

**Reply by arjunsinghrathore** [reply]
Yep, I realized that. I needed to do param.data

But just one doubt I still have is at what places can we access self.parameters(), as in I noticed that in training_step it does not return any params whereas in training_step_end for example I am able to get all the parameters of my model through self.parameters()

**Reply by rohitgr7** [reply]
it should be accessible everywhere.
